### PR TITLE
added logentries driver

### PR DESCRIPTION
**Pull request by caarlos0**
Add logentries log driver.

I used the logentries documentation and based my change uppon other log drivers implementations.

I'm not sure how to test this alone, this is my first contribution to docker as developer, so, any tips on this will be appreciated (a way to build and test only the logger will be awesome 👍 ).

### CODE REVIEW

**Comment by caarlos0** [comment]
(my commit is signed, don't know why it is complaining about it)

**Comment by cpuguy83** [comment]
@caarlos0 I don't see a signature in your commit.

**Comment by thaJeztah** [comment]
> my commit is signed, don't know why it is complaining about it

You signed it with GPG, but it needs a "signed-off-by .." in the commit message (`docker commit -s`). I usually do `docker commit -s -S`

**Comment by caarlos0** [comment]
@thaJeztah ohh, got it, fixed. Thanks =)

**Comment by StephenHynes7** [comment]
From a Logentries perspective everything looks fine, bsphere's lib is reliable and has been for a very long time. The only caveat though is that the logs do not send over SSL which is a bit sad tbh :(

**Comment by caarlos0** [comment]
@StephenHynes7 Yeah :(

**Comment by caarlos0** [comment]
@StephenHynes7 any lib that supports SSL?

**Comment by thaJeztah** [comment]
I'm not too familiar with logentries.com, @cpuguy83 told me it also implements syslog, so would there be downsides to using syslog instead of this driver?

Basically, we want to limit the amount of new logging drivers, and prefer to only add ones that are really necessary (until we have a system in place to allow external logging drives)

**Comment by caarlos0** [comment]
@thaJeztah It would require to have syslog installed, configured and running on the docker host...

I guess having the log driver in docker itself gives more power because the user can send logs to different accounts in different containers, etc...

**Comment by StephenHynes7** [comment]
> @StephenHynes7 any lib that supports SSL?

There is but I think the GO library is worth going forward with, I will look to see if I can create a PR for SSL option for it and we can see what happens.

**Comment by caarlos0** [comment]
@StephenHynes7 nice! Thanks

**Comment by StephenHynes7** [comment]
@caarlos0 fyi I have a PR to add tls support for the log library once that is merged then I would be happy [here](https://github.com/bsphere/le_go/pull/6).

**Comment by caarlos0** [comment]
@StephenHynes7 Nice, I'll take a look!

**Comment by StephenHynes7** [comment]
Okay so now this driver will send logs over TLS so as far as Logentries is concerned we are good 👍

**Comment by caarlos0** [comment]
@StephenHynes7 I'll update the submodule here

**Comment by thaJeztah** [comment]
@caarlos0 can you check the "vendor" failure to see if it's genuine? Looks like the vendor-check found a difference between the actually vendored files and the ones it's expecting;

```
13:38:39 The result of ./hack/vendor.sh differs
13:38:39 
13:38:39  D vendor/src/github.com/bsphere/le_go/.gitignore
13:38:39  D vendor/src/github.com/bsphere/le_go/.travis.yml
13:38:39  D vendor/src/github.com/bsphere/le_go/README.md
13:38:39  D vendor/src/github.com/bsphere/le_go/certs.go
13:38:39  D vendor/src/github.com/bsphere/le_go/le.go
13:38:39  D vendor/src/github.com/bsphere/le_go/le_test.go
```

**Comment by caarlos0** [comment]
@thaJeztah yeah, I need some help on that I guess..

I cloned the module inside the vendor folder and removed the `.git` folder, then added it to the `hack/vendor.sh`

Is that the correct way of doing this?

**Comment by thaJeztah** [comment]
Basically, you usually edit the script, then run it. This may take a while though (and I'm not sure it works correctly if you're running on OS X); alternatively you can fix by adding/removing the files mentioned above :-)

**Comment by caarlos0** [comment]
@thaJeztah If I run the script, basically all vendors change... don't know why. 

I'll try to comment the others and leave only le_go and then run it.

**Comment by caarlos0** [comment]
Nevermind, I got it, it removes all modules and clone them again...

running

**Comment by thaJeztah** [comment]
Yes, it can take a while; thinking of which, we recently merged a PR that allows updating just a single vendor; see the description at the top: https://github.com/docker/docker/blob/master/hack/vendor.sh#L8

**Comment by caarlos0** [comment]
should be good now :)

**Comment by caarlos0** [comment]
@thaJeztah the errors seems unrelated:

```
18:34:48 W: Failed to fetch http://llvm.org/apt/jessie/dists/llvm-toolchain-jessie-3.8/main/binary-amd64/Packages  404  Not Found
```

what can I do?

**Comment by thaJeztah** [comment]
@caarlos0 yes, we're aware of that issue; llvm.org today deleted/stopped their .deb repository, so we're looking for alternatives to get the builds going again

**Comment by caarlos0** [comment]
@thaJeztah Ah-ok.. let's wait then :)

thanks =)

**Comment by thaJeztah** [comment]
Darn, I suspect there's a bug in the vendor script; it seems so remove a file that's needed;

```
14:13:53 Wed Jun  1 14:13:53 UTC 2016
14:13:54 daemon/logger/logentries/logentries.go:9:2: cannot find package "github.com/bsphere/le_go" in any of:
14:13:54 	/go/src/github.com/docker/docker/vendor/github.com/bsphere/le_go (vendor tree)
14:13:54 	/usr/local/go/src/github.com/bsphere/le_go (from $GOROOT)
14:13:54 	/go/src/github.com/bsphere/le_go (from $GOPATH)
14:13:54 	/go/src/github.com/docker/docker/vendor/src/github.com/bsphere/le_go
```

@tianon any suggestions? ^^

**Comment by thaJeztah** [comment]
hm, I don't see any vendored files here now, did you let the script finish?

**Comment by thaJeztah** [comment]
ok, so running the script indeed removes those files :cry:

**Comment by thaJeztah** [comment]
Oh, erm, ah, looks like you forgot to register the log driver;
https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/daemon/logdrivers_windows.go#L6

and

https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/daemon/logdrivers_linux.go#L6

**Comment by caarlos0** [comment]
@thaJeztah I had... anyways, I re-added the files by hand...

Let's see what happens.

I also registered the logentries driver as requested.

**Comment by caarlos0** [comment]
Apparently running `./hack/vendor.sh git github.com/bsphere/le_go a56ef063a8d437513d430dfd9eaf9151ceace7ad` is the easiest way to get it right 👍

**Comment by thaJeztah** [comment]
That looks a lot better! Can you also squash your commits? Ideally, I think we should have two commits; one that brings in the "vendored" code (including the change to vendor.sh), and one that contains the changes in the docker code-base.

FYI; no decision has been made yet for this driver, but I'm adding this PR to our next review session for consideration; having the PR in an "almost ready" state could help moving forward after that (but no guarantees :blush:)

**Comment by caarlos0** [comment]
@thaJeztah sure, will do that :)

**Comment by caarlos0** [comment]
@thaJeztah should be good now. Let's see if the build passes =D

**Comment by thaJeztah** [comment]
Fingers crossed!

File: docs/admin/logging/logentries.md
Comment (thaJeztah) [inline docs/admin/logging/logentries.md]: note (for later); this line should be removed

File: docs/admin/logging/logentries.md
Comment (thaJeztah) [inline docs/admin/logging/logentries.md]: And this as well; that way it'll sort just alphabetically in the menu (but let's keep that change for later as well, during docs review)

Comment (caarlos0) [inline docs/admin/logging/logentries.md]: ok :)

File: docs/admin/logging/logentries.md
Comment (caarlos0) [inline docs/admin/logging/logentries.md]: ok :)

**Comment by thaJeztah** [comment]
We discussed this in the maintainers session; and we're good with including this, as it's a small dependency so should be doable to maintain :smile:

**Comment by caarlos0** [comment]
@thaJeztah NICE =D

What are the next steps?

File: vendor/src/github.com/bsphere/le_go/certs.go
Comment (cpuguy83) [inline vendor/src/github.com/bsphere/le_go/certs.go]: Using hard-coded certs scares me a little.
It would be better to rely on the host's certificate store.
As is, if the logentries cert was ever compromised it would require releasing a new version of Docker to fix.

**Comment by thaJeztah** [comment]
@caarlos0 it's code-review time now, and I see @cpuguy83 has started with commenting 👍

File: vendor/src/github.com/bsphere/le_go/certs.go
Comment (caarlos0) [inline vendor/src/github.com/bsphere/le_go/certs.go]: @cpuguy83 makes sense... I guess I could try to open a PR to le_go to fix that...

Comment (caarlos0) [inline vendor/src/github.com/bsphere/le_go/certs.go]: @cpuguy83 well, I did some research and didn't found a way of doing it without the certificates...

what I found is that there were too many certificates there, so I removed those which were not needed. https://github.com/bsphere/le_go/pull/10

Any suggestions?

**Comment by caarlos0** [comment]
@thaJeztah 😄

**Comment by thaJeztah** [comment]
ping @cpuguy83

File: vendor/src/github.com/bsphere/le_go/certs.go
Comment (StephenHynes7) [inline vendor/src/github.com/bsphere/le_go/certs.go]: I presume we agreeing that this is still a blocker?

Comment (caarlos0) [inline vendor/src/github.com/bsphere/le_go/certs.go]: @StephenHynes7 don't know, I need some suggestions on how to fix this.

BTW: I don't think it is a common thing to change the certificate, and, I presume, other log drivers running over HTTPS will also need the certificate somehow, am I right?

**Comment by caarlos0** [comment]
ping 😞

**Comment by StephenHynes7** [comment]
We all good? Everything is green?

**Comment by thaJeztah** [comment]
I agree with @cpuguy83 that having a hard-coded certificate is an issue; we can't do a new release if the certificate is compromised, especially since we don't do back ports to older versions of docker. In such a situation, every user that's running an older version of docker (some people can't upgrade to newer versions for many reasons, or by policy).

I think the certificate should be external, and the installation/configuration instructions for this driver should include how to obtain, and install the certificate on the Docker host.

So, yes, that's a blocker

**Comment by caarlos0** [comment]
@thaJeztah understood... that would require more changes in the logentries dependency...

would it be a problem if the dep still has the keys in it, but also had a way to connect passing the PATH to the keys, and we use only the last one?

For people who use logentries already, it might be bad to just broke the api... so I thought we could just add one more entrypoint and use it in docker....


let me know your thoughts...

**Comment by thaJeztah** [comment]
Perhaps that would be acceptable, as long as we don't use the built-in certificate by default (i.e., that we *require* people to install an external certificate). That's just my thought, so happy to hear @cpuguy83's opinion 👍

File: vendor/src/github.com/bsphere/le_go/certs.go
Comment (NathanMcCauley) [inline vendor/src/github.com/bsphere/le_go/certs.go]: These certs are issued by a trusted CA (geotrust). Why do they need to be embedded or configured at all? Are you trying to do certificate pinning here?

Comment (endophage) [inline vendor/src/github.com/bsphere/le_go/certs.go]: @NathanMcCauley would it require upgrading docker to Go 1.7 to correctly read the system certs via the new `x509.SystemCertPool()` function?

Comment (caarlos0) [inline vendor/src/github.com/bsphere/le_go/certs.go]: @NathanMcCauley don't know, they are in the logentries library...

Comment (cpuguy83) [inline vendor/src/github.com/bsphere/le_go/certs.go]: But wasn't this added at our behest at not wanting to use a non-encrypted transport, meanwhile at the time the library only supported HTTP, not HTTPS.

**Comment by cpuguy83** [comment]
Definitely no problem with adding this driver.  Library is clean and super tiny, but we really need to follow good practices and not embed the certificate in docker's code base.

File: vendor/src/github.com/bsphere/le_go/certs.go
Comment (caarlos0) [inline vendor/src/github.com/bsphere/le_go/certs.go]: @cpuguy83 yes, it was added by @StephenHynes7 ... maybe we can just use system cert pool, I'll try to do that.

Comment (caarlos0) [inline vendor/src/github.com/bsphere/le_go/certs.go]: @cpuguy83 @endophage @NathanMcCauley using `SystemCertPool()` works, but will require Go 1.7.

Seems like docker is already being compiled under Go 1.7, so, I guess this is alright...

**Comment by caarlos0** [comment]
I just opened a PR removing the certificates from the le_go lib: bsphere/le_go#12

**Comment by caarlos0** [comment]
Updated the le_go submodule, vendored and rebased commits with master.

**Comment by cpuguy83** [comment]
LGTM

**Comment by cpuguy83** [comment]
Oh except vendor is not quite right:
```
19:04:42 The result of ./hack/vendor.sh differs
19:04:42 
19:04:42  D vendor/src/github.com/bsphere/le_go/le_test.go
19:04:42 
19:04:42 Please vendor your package with ./hack/vendor.sh.
```

**Comment by caarlos0** [comment]
@cpuguy83 oops

**Comment by caarlos0** [comment]
@cpuguy83 should be good now

**Comment by cpuguy83** [comment]
re-LGTM.

**Comment by caarlos0** [comment]
ping?

**Comment by mlaventure** [comment]
LGTM.

Moving to doc review.

re-ping @thaJeztah

File: docs/admin/logging/logentries.md
Comment (thaJeztah) [inline docs/admin/logging/logentries.md]: This document was moved to the central docs repository. Given that will be part of 1.13, and should not be published yet, can you remove this file here, and open a PR against the `vnext-engine` branch in that repository?

https://github.com/docker/docker.github.io/tree/vnext-engine/engine/admin/logging

**Review by thaJeztah — CHANGES_REQUESTED** [review]
I don't think this logging driver supports viewing logs through `docker logs` (correct?) in that case no changes are needed in the reference docs. The documentation in this PR should go into the central docs repository though (see my comment).

Also, we need some changes in the bash and zsh completion scripts; https://github.com/docker/docker/blob/f08a450ad35706900766e7766d30a90d688e410b/contrib/completion/bash/docker#L499-L561 https://github.com/docker/docker/blob/6b997739fe5f4de411ccf30be574ee11a2780018/contrib/completion/zsh/_docker#L217-L251

Let us know if you want to try making those changes, or rather leave that to @albers and @sdurrheimer (who maintain those scripts)

**Comment by caarlos0** [comment]
@thaJeztah 

Seems like I deleted my fork and github lost track of the PR.

I tried re-create the fork and the branch from the pull request ref (`refs/pull/22699/head`) and push it again, but it doesn't seem to be working.

Should I open another pr?

**Comment by thaJeztah** [comment]
@caarlos0 yes, probably not possible to update the PR if you had to create a new branch. Let me know when it's up then we can link to it, and close this one

**Comment by caarlos0** [comment]
@thaJeztah I just created a PR for the docs in https://github.com/docker/docker.github.io/pull/244

**Comment by caarlos0** [comment]
@thaJeztah #27471. Can I close this one?

**Comment by thaJeztah** [comment]
Yes, let's close, thanks!

**Pull request closed without merging**
